### PR TITLE
Order invariant loss

### PR DESCRIPTION
**Pull request by TorbenSDJohansen**
This PR adds the proposed block order-invariant loss module (`OrderInvariantSeq2SeqCrossEntropy`) for our new generation OccCANINE model. See the documentation in [histocc.loss.order_invariant.py](https://github.com/christianvedels/OccCANINE/blob/order-invariant-loss/histocc/loss/order_invariant.py) for details, as well as its associated [unit tests](https://github.com/christianvedels/OccCANINE/blob/order-invariant-loss/tests/test_loss.py).

The proposed module involves a fair number of moving pieces, and a close look by a second person is recommended.

For completeness, relevant parts of `OrderInvariantSeq2SeqCrossEntropy`'s documentation are listed below:

> Implementation of loss function for seq2seq model which is invariant to the order of which "blocks" (i.e., HISCO codes) are predicted. The motivation is that we do not particularly care whether a model predicts whether an occupational description corresponds to (1) Farmer & fisher or (2) Fisher & farmer. Further, we cannot guarantee that our labels adhere to some well-
defined structure which makes the order of labels inferable from the occupational description.

Its class documentation:
> seq2seq-style loss function for sequences consisting of `nb_blocks` blocks, all of size `block_size`. The module consists of two types of losses: 
1) BLOCK ORDER-INVARIANT CLASSIFICATION: 
For each target block, calculate the loss with respect to all `nb_blocks` candidate blocks, using ordinary cross entropy as the loss function. Then choose the smallest of the `nb_blocks` candidate losses as the loss with respect to the given target block. Do so for each of the `nb_blocks` target blocks and average these to obtain the final loss related to the block order-invariant classification loss.
Note that the above procedure is calculated separately for each observation in the input batch, i.e., for the first target block, two different observations in the same batch may end up selecting different candidate blocks.
Additionally, since most targets consist mainly of empty blocks, which are denoted using the `pad_idx` value, we ignore its index when calculating the loss. This ensures that the loss for the second target block only depends on the observations in the batch for which the second target block is not empty.
2) PUSH TOWARDS SPARSITY
To ensure the model is pushed towards only producing as many candidates as there are actual non-empty target blocks, apply cross entropy with respect to the entire sequence, using `pad_idx` as the target value.

### CODE REVIEW

File: histocc/loss/order_invariant.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+        for target_block in range(self.nb_blocks):
+            # Look at target block and calculate loss
+            # with respect to all candidate predictions
+
+            start_idx = target_block * self.block_size
+            end_idx = start_idx + self.block_size
+
+            if (target[:, start_idx:end_idx] == self.pad_idx).all():
+                break # Only padding remains, which we ignore
```
Comment (christianvedels) [inline histocc/loss/order_invariant.py:133]: Smart

File: histocc/loss/order_invariant.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        Tensor
+            Total loss stored as a 1-element tensor.
+        '''
+        yhat = yhat.permute(0, 2, 1)[:, :, :-1] # -> [BATCH_SIZE, VOCAB, BLOCK_SIZE * NB_BLOCKS]
+        target = target[:, 1:-1] # ignore initial BOS and final EOS
+
+        order_invariant_loss = self._order_invariant_loss(yhat, target)
+        push_to_pad_loss = self._push_to_pad(yhat)
+
+        loss = order_invariant_loss + self.push_to_pad_scale_factor * push_to_pad_loss
```
Comment (christianvedels) [inline histocc/loss/order_invariant.py:185]: Out of curiosity: Why are the losses added together and not multiplied?

**Comment by christianvedels** [comment]
I can't find any issues here. Minor comments are included above.
